src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs

```diff
@@ -208,6 +208,104 @@ internal sealed override IEnumerable<Cci.INestedTypeDefinition> GetSynthesizedNe
             return null;
         }
 
+        public sealed override ImmutableArray<(Cci.ITypeDefinition, ImmutableArray<Cci.DebugSourceDocument>)> GetTypeToDebugDocumentMap(EmitContext context)
+        {
+            var debugDocuments = ArrayBuilder<Cci.DebugSourceDocument>.GetInstance();
+            var methodDocumentList = PooledHashSet<Cci.DebugSourceDocument>.GetInstance();
+            var result = ArrayBuilder<(Cci.ITypeDefinition, ImmutableArray<Cci.DebugSourceDocument>)>.GetInstance();
+
+            var namespacesAndTypesToProcess = ArrayBuilder<NamespaceOrTypeSymbol>.GetInstance();
+            namespacesAndTypesToProcess.Push(SourceModule.GlobalNamespace);
+            while (namespacesAndTypesToProcess.Count > 0)
+            {
+                var symbol = namespacesAndTypesToProcess.Pop();
+
+                switch (symbol.Kind)
+                {
+                    case SymbolKind.Namespace:
+                        var location = GetSmallestSourceLocationOrNull(symbol);
+
+                        // filtering out synthesized symbols not having real source 
+                        // locations such as anonymous types, etc...
+                        if (location != null)
+                        {
+                            foreach (var member in symbol.GetMembers())
+                            {
+                                switch (member.Kind)
+                                {
+                                    case SymbolKind.Namespace:
+                                    case SymbolKind.NamedType:
+                                        namespacesAndTypesToProcess.Push((NamespaceOrTypeSymbol)member);
+                                        break;
+                                    default:
+                                        throw ExceptionUtilities.UnexpectedValue(member.Kind);
+                                }
+                            }
+                        }
+                        break;
+                    case SymbolKind.NamedType:
+                        var typeDefinition = (Cci.ITypeDefinition)symbol.GetCciAdapter();
+                        GetDocumentsForMethods(methodDocumentList, typeDefinition, context);
```

Comment:
>GetDocumentsForMethods [](http://example.com/codeflow?start=24&length=22)

Do we need to get the set of documents containing methods rather than just looking at the documents associated with `symbol.Locations` below? Are there cases where methods are not contained in the spans from `symbol.Locations`?

Reply:
The idea is to not duplicate any information in the PDB, so if we're going to write out a document for a method, then we don't want to write out the same document for the type that contains that method.